pkg/scheduler/ugm/queue_tracker.go

```diff
@@ -245,34 +245,32 @@ func (qt *QueueTracker) setLimit(queuePath string, maxResource *resources.Resour
 	childQueueTracker.maxResources = maxResource
 }
 
-func (qt *QueueTracker) headroom(queuePath string) *resources.Resource {
+func (qt *QueueTracker) headroom(hierarchy []string) *resources.Resource {
 	log.Log(log.SchedUGM).Debug("Calculating headroom",
-		zap.String("queue path", queuePath))
-	childQueuePath, immediateChildQueueName := getChildQueuePath(queuePath)
-	if childQueuePath != common.Empty {
-		if qt.childQueueTrackers[immediateChildQueueName] != nil {
-			headroom := qt.childQueueTrackers[immediateChildQueueName].headroom(childQueuePath)
-			if headroom != nil {
-				return resources.ComponentWiseMinPermissive(headroom, qt.maxResources)
-			}
-		} else {
-			log.Log(log.SchedUGM).Error("Child queueTracker tracker must be available in child queues map",
-				zap.String("child queueTracker name", immediateChildQueueName))
-			return nil
+		zap.Strings("queue path", hierarchy))
+	// depth first: all the way to the leaf, create if not exists
+	// more than 1 in the slice means we need to recurse down
+	var headroom, childHeadroom *resources.Resource
+	if len(hierarchy) > 1 {
+		childName := hierarchy[1]
+		if qt.childQueueTrackers[childName] == nil {
+			qt.childQueueTrackers[childName] = newQueueTracker(qt.queuePath, childName)
 		}
+		childHeadroom = qt.childQueueTrackers[childName].headroom(hierarchy[1:])
 	}
-
+	// arrived at the leaf or on the way out: check against current max if set
 	if !resources.Equals(resources.NewResource(), qt.maxResources) {
-		headroom := qt.maxResources.Clone()
+		headroom = qt.maxResources.Clone()
 		headroom.SubOnlyExisting(qt.resourceUsage)
 		log.Log(log.SchedUGM).Debug("Calculated headroom",
-			zap.String("queue path", queuePath),
-			zap.String("queue", qt.queueName),
+			zap.String("queue path", qt.queuePath),
 			zap.String("max resource", qt.maxResources.String()),
 			zap.String("headroom", headroom.String()))
-		return headroom
 	}
-	return nil
+	if headroom == nil {
+		return childHeadroom
+	}
+	return resources.ComponentWiseMinPermissive(headroom, childHeadroom)
```

Comment:
I tried to reproduce the issue by using the modified test method and thought of simplifying the changes and also to retain the same signature like given below:

mani@mani-MBP16 incubator-yunikorn-core % git diff pkg/scheduler/ugm/queue_tracker.go
diff --git a/pkg/scheduler/ugm/queue_tracker.go b/pkg/scheduler/ugm/queue_tracker.go
index c793ffc..1b089de 100644
--- a/pkg/scheduler/ugm/queue_tracker.go
+++ b/pkg/scheduler/ugm/queue_tracker.go
@@ -253,7 +253,11 @@ func (qt *QueueTracker) headroom(queuePath string) *resources.Resource {
                if qt.childQueueTrackers[immediateChildQueueName] != nil {
                        headroom := qt.childQueueTrackers[immediateChildQueueName].headroom(childQueuePath)
                        if headroom != nil {
-                               return resources.ComponentWiseMinPermissive(headroom, qt.maxResources)
+                               pHeadroom := qt.maxResources.Clone()
+                               if !resources.Equals(resources.NewResource(), pHeadroom) {
+                                       pHeadroom.SubOnlyExisting(qt.resourceUsage)
+                               }
+                               return resources.ComponentWiseMinPermissive(headroom, pHeadroom)
                        }
                } else {
                        log.Log(log.SchedUGM).Error("Child queueTracker tracker must be available in child queues map",

Please assess this alternative approach.